src/coins.cpp

```diff
@@ -267,17 +267,19 @@ bool CCoinsViewCache::Flush() {
 
 bool CCoinsViewCache::Sync()
 {
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, /*will_erase=*/false);
-    // Instead of clearing `cacheCoins` as we would in Flush(), just clear the
-    // FRESH/DIRTY flags of any coin that isn't spent.
-    for (auto it = cacheCoins.begin(); it != cacheCoins.end(); ) {
+    bool fOk = base->BatchWrite(m_flagged_head.Next(), hashBlock, /*will_erase=*/false);
+    // Instead of clearing `cacheCoins` as we would in Flush(), just erase the
+    // spent coins. All coins left in the linked list after BatchWrite should be spent.
+    for (auto it{m_flagged_head.Next()}; it != nullptr; ) {
+        const auto next_entry{it->second.Next(/*clear_flags=*/true)};
         if (it->second.coin.IsSpent()) {
             cachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
-            it = cacheCoins.erase(it);
-        } else {
-            it->second.ClearFlags();
-            ++it;
+            cacheCoins.erase(it->first);
+        } else if (fOk) {
```

Comment:
Neat! Could also add `[[unlikely]]` here https://github.com/bitcoin/bitcoin/pull/28280/files#diff-f0ed73d62dae6ca28ebd3045e5fc0d5d02eaaacadb4c2a292985a3fbd7e1c77cR189

Reply:
I think I will refrain from adding these annotations for now, since they are more than just documentation and may have side effects in the generated code. I might experiment with a separate commit to see if they affect performance.